src/back/auth/src/service/service.cpp

```diff
@@ -156,6 +163,70 @@ model::UserInfo Service::GetUserInfo(const std::string& sessionId)
 	return _oidc.GetUserInfo(session._accessToken);
 }
 
+model::SessionRefresh Service::RefreshSession(
+	const std::string& sessionId,
+	const std::string& userAgent)
+{
+	// запрашиваем сессию, не важно, активна она или нет
+	auto session = _session.Table().Get(sessionId);
+
+	// проверка на несовпадение userAgent
+	differentDeviceSecurityCheck(userAgent, session._device);
+	
+	// проверка на неактивность сессии
+	if (session._active == false){
+		_session.Table().BlockEverySessionByUser(session._userId);
+		throw errors::SecurityRisk{"Same inactive session."};
+	}
+		
+	// Обновляем OIDC токены
+	updateTokens(session);		
+	
+	// Обновляем токен Сессии
+	std::string token;
+	try {
+		token = updateSession(session, userAgent);
+	} catch(errors::SecurityRisk& e){
+		_session.Table().BlockEverySessionByUser(session._userId);
+		throw;
+	}
+	
+	model::SessionRefresh refresh = {
+		._token = token
+	};
+
+	return refresh;
+}
+
+std::string Service::updateSession(
+	const model::Session& session,
+	const std::string& userAgent){
+
+	const OIDCTokens tokens = {
+		._accessToken = session._accessToken,
+		._refreshToken = session._refreshToken,
+		._idToken = session._idToken
+	};
+	
+	const auto data = _tokenizer.OIDC().Parse(tokens._accessToken);
+
+	const auto exp = _tokenizer.GetExpirationTime(tokens._refreshToken);
+
+	const auto newSession = _session.Refresh(tokens, data, userAgent, exp, session._id);
+
+	return newSession._token;
+
+}
+
+void Service::differentDeviceSecurityCheck(
+	const std::string& currentUserAgent,
+	const std::string& oldUserAgent
+){
+	if (currentUserAgent != oldUserAgent) {
+		throw errors::SecurityRisk{"Different user agent."};
```

Comment:
Нужно бы залогировать агент.
```suggestion
		const auto agent = currentUserAgent.substr(0, 150);
		throw errors::SecurityRisk{"Different user agent: '" + agent + '\''};
```

Reply:
Готово.